test/mocks/upstream/host.h

```diff
@@ -46,7 +46,7 @@ class MockDetector : public Detector {
   MockDetector();
   ~MockDetector();
 
-  void runCallbacks(HostSharedPtr host) {
+  void runCallbacks(const HostSharedPtr& host) {
```

Comment:
This is the troublesome line.  By using a reference here the "host" variable refers back to what's passed into runCallbacks in sds_test.cc:244 rather than making a shared_ptr copy and thus reference count increase.  One of the callbacks modifies the host list returned by cluster_->hosts(), with this change the reference count indicates it is safe to free the host when the host list is updated.  The callback itself DOES use pass-by-value function argument, so when it returns the HostSharedPtr goes out of scope resulting in a read of the reference count (but the object has already been deleted).

It can be const, but not a reference.